Models/MuZero_agent_2.py

```diff
@@ -584,88 +589,136 @@ def player_to_play(player_num, action):
 
     @staticmethod
     def visit_softmax_temperature():
-        return .1
+        return 1
 
 
-class Batch_MCTSAgent(MCTSAgent):
-    """
-    Use Monte-Carlo Tree-Search to select moves.
-    """
+class MCTS(MCTSAgent):
+    def __init__(self, network):
+        super().__init__(network, 0)
+        self.times = [0]*6
 
+    
     def __init__(self, network: Network) -> None:
         super().__init__(network, 0)
         self.NUM_ALIVE = config.NUM_PLAYERS
-
-    # Core Monte Carlo Tree Search algorithm.
-    # To decide on an action, we run N simulations, always starting at the root of
-    # the search tree and traversing the tree according to the UCB formula until we
-    # reach a leaf node.
-    def run_batch_mcts(self, root: list, action: List):
-        min_max_stats = [MinMaxStats(config.MINIMUM_REWARD, config.MAXIMUM_REWARD) for _ in range(config.NUM_PLAYERS)]
-        
-        for _ in range(config.NUM_SIMULATIONS):
-            history = [ActionHistory(action[i]) for i in range(self.NUM_ALIVE)]
-            node = root
-            search_path = [[node[i]] for i in range(self.NUM_ALIVE)]
-            
-            # There is a chance I am supposed to check if the tree for the non-main-branch
-            # Decision paths (axis 1-4) should be expanded. I am currently only expanding on the
-            # main decision axis.
-            for i in range(self.NUM_ALIVE):
-                while node[i].expanded():
-                    action[i], node[i] = self.select_child(node[i], min_max_stats[i])
-                    history[i].add_action(action[i])
-                    search_path[i].append(node[i])
-            
+    
+    def run_batch_mcts(self, roots_cpp, hidden_state_pool):
+        # preparation
+        num = roots_cpp.num
+        # config variables
+        discount = config.DISCOUNT  
+        pb_c_init = config.PB_C_INIT
+        pb_c_base = config.PB_C_BASE
+        hidden_state_index_x = 0
+
+
+        # minimax value storage data structure 
+        min_max_stats_lst = tree.MinMaxStatsList(num)
+        min_max_stats_lst.set_delta(config.MAXIMUM_REWARD*2)  # config.MINIMUM_REWARD *2 
+        horizons = 1  # self.config.lstm_horizon_len, seems to be the number of timesteps predicted in the future 
+        hidden_state_pool = [hidden_state_pool]
+        for _ in range(config.NUM_SIMULATIONS): # go through the tree NUM_SIMULATIONS times 
+            # prepare a result wrapper to transport results between python and c++ parts
+            hidden_states = [] 
+            results = tree.ResultsWrapper(num)
+
+            # evaluation for leaf nodes, traversing across the tree and updating values
+            hidden_state_index_x_lst, hidden_state_index_y_lst, last_action = \
```

Comment:
Can you explain why we are returning what looks to be 2 tensors. Why are we not returning a single tensor which is the batch of hidden_states.

Reply:
I think that list x references the player, and list y references the timestep. These lists are then used to collect up the relevant data to be fed to the model. I suppose that another approach would be to just combine x and y lists into one, but the efficient zero code does it like this and by changing it we risk breaking something accidentally which is hard to detect, for what seems like very little benefit other than the code being slightly neater